File: day_2/part_2.py
def is_invalid_id(n: int) -> bool:
    s = str(n)
    length = len(s)

    if s[0] == "0":
        return False

    for pattern_len in range(1, length // 2 + 1):
        if length % pattern_len == 0:
            pattern = s[:pattern_len]
            repeats = length // pattern_len

            if pattern * repeats == s and repeats >= 2:
                return True

    return False


def solve(input: str) -> int:
    ranges = []
    for range_str in input.strip().split(","):
        range_str = range_str.strip()
        if range_str:
            start, end = map(int, range_str.split("-"))
            ranges.append((start, end))

    sum = 0

    for start, end in ranges:
        for i in range(start, end + 1):
            if is_invalid_id(i):
                sum += i

    return sum


if __name__ == "__main__":
    with open("day_2.txt", "r") as f:
        input = f.read()
        result = solve(input)
        print(result)

File: go/day_3/part_2.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func solve(input string) int {
	banks := []string{}
	for _, line := range strings.Split(strings.TrimSpace(input), "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			banks = append(banks, line)
		}
	}

	sum := 0

	for _, b := range banks {
		n := len(b)
		k := 12
		result := []rune{}
		start := 0

		for i := 0; i < k; i++ {
			end := n - k + i
			max_digit := rune('0')
			for j := start; j <= end; j++ {
				if b[j] > byte(max_digit) {
					max_digit = rune(b[j])
				}
			}

			for j := start; j <= end; j++ {
				if rune(b[j]) == max_digit {
					result = append(result, max_digit)
					start = j + 1
					break
				}
			}
		}

		joltage, _ := strconv.Atoi(string(result))
		sum += joltage
	}

	return sum
}

func main() {
	input, err := os.ReadFile("../inputs/day_3.txt")
	if err != nil {
		panic(err)
	}
	result := solve(string(input))
	fmt.Println(result)
}
